graviti/openapi/policy.py

```diff
@@ -5,9 +5,101 @@
 
 """Interfaces about the dataset policy."""
 
-from typing import Any, Dict, Optional
+import base64
+import hashlib
+import hmac
+from datetime import datetime, timezone
+from typing import Any, ByteString, Dict, Optional, Tuple
 
-from graviti.openapi.requests import open_api_do
+import xmltodict
+
+from graviti.exception import ResponseError
+from graviti.openapi.requests import do, open_api_do
+from graviti.utility import config, convert_iso_to_datetime, locked
+
+_EXPIRED_IN_SECOND = 240
+_GMT_FORMAT = "%a, %d %b %Y %H:%M:%S GMT"
+_PERMISSIONS: Dict[Tuple[str, str, str, str], Dict[str, Any]] = {}
+_DEFAULT_PERMISSION = {"expireAt": datetime.fromtimestamp(0, timezone.utc)}
+
+
+@locked
+def _request_read_permission(
+    access_key: str,
+    url: str,
+    owner: str,
+    dataset: str,
+) -> None:
+    permission: Dict[str, Any] = get_object_policy(
+        access_key,
+        url,
+        owner,
+        dataset,
+        is_internal=config.is_internal,
+        expired=_EXPIRED_IN_SECOND,
+    )
+    permission["expireAt"] = convert_iso_to_datetime(permission["policy"].pop("expireAt"))
+    _PERMISSIONS[access_key, owner, dataset, "GET"] = permission
+
+
+def _get_read_permission(
+    access_key: str,
+    url: str,
+    owner: str,
+    dataset: str,
+    request_again: bool,
+) -> Dict[str, Any]:
+    key = (access_key, owner, dataset, "GET")
+    if (
+        request_again
+        or datetime.now(timezone.utc) >= _PERMISSIONS.get(key, _DEFAULT_PERMISSION)["expireAt"]
+    ):
+        _request_read_permission(
+            access_key,
+            url,
+            owner,
+            dataset,
+        )
+
+    return _PERMISSIONS[key]
+
+
+def _get_oss_authorization(  # pylint: disable=too-many-arguments
+    access_key_id: str,
+    access_key_secret: str,
+    security_token: str,
+    bucket: str,
+    object_name: str,
+    verb: str,
+    date: str,
+) -> str:
+    signature = (
+        f"{verb}\n\n\n{date}\nx-oss-security-token:{security_token}\n/{bucket}/{object_name}"
+    )
+    hashing = hmac.new(access_key_secret.encode("utf-8"), signature.encode("utf-8"), hashlib.sha1)
+    signature = base64.b64encode(hashing.digest()).decode("utf-8")
+    return f"OSS {access_key_id}:{signature}"
+
+
+def _get_binary_file_from_oss(  # pylint: disable=too-many-arguments
+    access_key_id: str,
+    access_key_secret: str,
+    security_token: str,
+    endpoint: str,
+    bucket: str,
+    object_name: str,
+) -> ByteString:
```

Comment:
why the return type is ByteString, can we get a FilePointer like object from OSS and use read to lazy load the binary content?